### DISCUSSION TITLE
Lesson 7, Part 19, Final Test Fails

### DISCUSSION THREAD

**Discussion by ultraKelevra** [Cyfrin/foundry-full-course-cu#1352] [Q&A] [answered]
### Tests for fund and withdraw in the last segment of Lesson 7, Part 19 not working for me

I made it all the way to the end to this part. But my tests fail when trying to run the Interactions tests. The tests in the video run perfectly but my code is almost the same and still fails. I've noticed that in the video, Patrick does not need to `prank `any user for his integration tests to pass. I've tried pranking and then dealing some ETH to the `USER` account but still the only way the `fund()` test works is by dealing ETH to the `fundFundMe` contract. Is logical, as that's the contract that is executing the interaction. But I don't understand why pranking the user right before, doesn't overwrite the transaction sender. Anyway, I managed to make the test pass by just funding the funder contract as the user.

Then, when testing `withdraw()` it failed again. So I tried `startPrank(fundMe.getUser())` just like in the unit test, but it still reverts with `[FAIL. Reason: NotOwner()]` I've logged the addresses of fundFundMe, withdrawFundMe, deployFundMe, and `address(this)` but none of them is the owner of fundMe.

So basically now I'm confused. I've followed the videos and everything was hard but rewarding. In the end all my unit tests were green so I know is not a problem in the FundMe contract itself. Now I feel like I don't understand what is going on.

Copilot Chat isn't helping either.

I think I've come to the end of what I can do by myself. Please help

### Interactions.s.sol
```
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

import {Script, console} from "forge-std/Script.sol";
import {DevOpsTools} from "lib/foundry-devops/src/DevOpsTools.sol";
import {FundMe} from "../src/FundMe.sol";

contract FundFundMe is Script {
    uint256 constant SEND_VALUE = 0.01 ether;

    function fundFundMe(address mostRecentlyDeployed) public {
        FundMe(payable(mostRecentlyDeployed)).fund{value: SEND_VALUE}();
        console.log("Funded FundMe contract with: %s", SEND_VALUE);
    }

    function run() external {
        address mostRecentlyDeployed = DevOpsTools.get_most_recent_deployment("FundMe", block.chainid);

        vm.startBroadcast();
        fundFundMe(mostRecentlyDeployed);
        vm.stopBroadcast();
    }
}

contract WithdrawFundMe is Script {
    uint256 constant SEND_VALUE = 0.01 ether;

    function withdrawFundMe(address mostRecentlyDeployed) public {
        FundMe(payable(mostRecentlyDeployed)).withdraw();
        console.log("Withdrawn FundMe contract with: %s", SEND_VALUE);
    }

    function run() external {
        address mostRecentlyDeployed = DevOpsTools.get_most_recent_deployment("FundMe", block.chainid);

        vm.startBroadcast();
        withdrawFundMe(mostRecentlyDeployed);
        vm.stopBroadcast();
    }
}
```

### InteractionsTest.t.sol
```
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

import {Test, console} from "forge-std/Test.sol";
import {FundMe} from "../../src/FundMe.sol";
import {DeployFundMe} from "../../script/DeployFundMe.s.sol";
import {FundFundMe, WithdrawFundMe} from "../../script/Interactions.s.sol";

contract InteractionsTest is Test {
    FundMe fundMe;
    address USER = makeAddr("user");
    uint256 SEND_VALUE = 0.01 ether;
    uint256 STARTING_BALANCE = 100 ether;

    function setUp() external {
        DeployFundMe deployFundMe = new DeployFundMe();
        fundMe = deployFundMe.run();
    }

    function testUserCanFundInteractions() public {
        // arrange
        FundFundMe fundFundMe = new FundFundMe();
        WithdrawFundMe withdrawFundMe = new WithdrawFundMe();

        // act
        // --fund
        vm.deal(address(fundFundMe), STARTING_BALANCE);
        fundFundMe.fundFundMe(address(fundMe));

        // --withdraw
        vm.startPrank(fundMe.getOwner());
        withdrawFundMe.withdrawFundMe(address(fundMe));
        vm.stopPrank();

        // assert
        assert(address(fundMe).balance == 0);
    }
}

```

**Comment by Blockchain-Oracle** [comment, marked as answer]
If here is the error you encounted
```
Running 1 test for test/interactions.t.sol:FundMeTest
[FAIL. Reason: you have an active prank; broadcasting and pranks are not compatible] testUserFundInteractions() (gas: 215170)
Test result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 2.03ms
 
Ran 1 test suites: 0 tests passed, 1 failed, 0 skipped (1 total tests)
```


**Heres the solution comment out prank , 
once you broadcast pranks dont work!**

```solidity

  // --withdraw
        // vm.startPrank(fundMe.getOwner());
        withdrawFundMe.fundFundMe(address(fundMe));
        // vm.stopPrank();
       
```

and in your interaction, here is the correct syntax

```solidity
 function withdrawFundMe(address mostRecentlyDeployed) public {
        vm.startBroadcast();

        FundMe(payable(mostRecentlyDeployed)).withdraw();
        vm.stopBroadcast();

        console.log("Withdrawn FundMe contract with: %s", SEND_VALUE);
    }
    
```

```solidity 
contract WithdrawFundMe is Script {
    uint256 constant SEND_VALUE = 0.3 ether;

    function withdrawFundMe(address mostRecentlyDeployed) public {
        vm.startBroadcast();

        FundMe(payable(mostRecentlyDeployed)).withdraw();
        vm.stopBroadcast();

        console.log("Withdrawn FundMe contract with: %s", SEND_VALUE);
    }

    function run() external {
        address mostRecentlyDeployed = DevOpsTools.get_most_recent_deployment(
            "FundMe",
            block.chainid
        );

        withdrawFundMe(mostRecentlyDeployed);
    }
}
```


### remeber to mark this as answer!

**Reply by ultraKelevra** [reply]
Thanks It worked!

But why is the `withdrawFundMe()` function using broadcast in the function instead of in `run()`, while `FundFundMe` contract has it the other way around? When the test first failed in the video, Patrick said that broadcast should be "in the function below" (`run()`) I assumed that from now on, in all my interactions scripts, the broadcast should be in the run and leave the function simple. I know I'll get this with practice and testing, but it would be great if you explain to me that last bit. Just a simple answer, don't need a lot. Just a reason I can later google and such. Thanks sincerely.

**Reply by Blockchain-Oracle** [reply]
Yh i totally agree with you!. but you were caling this function directly in your test 

```solidity
fundFundMe.fundFundMe(address(fundMe));
 
```
**and putting broadcast in run**

```solidity

    function run() external {
        address mostRecentlyDeployed = DevOpsTools.get_most_recent_deployment("FundMe", block.chainid);

        vm.startBroadcast();
        fundFundMe(mostRecentlyDeployed);
        vm.stopBroadcast();
    }
```

**leaving this function behind with no  broadcast**

```solidity
   function fundFundMe(address mostRecentlyDeployed) public {
        FundMe(payable(mostRecentlyDeployed)).fund{value: SEND_VALUE}();
        console.log("Funded FundMe contract with: %s", SEND_VALUE);
    }
```
